### ISSUE TITLE
Add wrappers for Nevergrad

### ISSUE THREAD

**Issue by janosg** [optimagic-dev/optimagic#560]
Wrap the gradient free optimizers from [nevergrad](https://github.com/facebookresearch/nevergrad) in optimagic.

Nevergrad implements the following algorithms:

- `NgIohTuned` is "meta"-optimizer which adapts to the provided settings (budget, number of workers, parametrization) and should therefore be a good default.
- `TwoPointsDE` is excellent in many cases, including very high `num_workers`.
- `PortfolioDiscreteOnePlusOne` is excellent in discrete settings of mixed settings when high precision on parameters is not relevant; it's possibly a good choice for hyperparameter choice.
- `OnePlusOne` is a simple robust method for continuous parameters with `num_workers` < 8.
- `CMA` is excellent for control (e.g. neurocontrol) when the environment is not very noisy (num_workers ~50 ok) and when the budget is large (e.g. 1000 x the dimension).
- `TBPSA` is excellent for problems corrupted by noise, in particular overparameterized (neural) ones; very high `num_workers` ok).
- `PSO` is excellent in terms of robustness, high `num_workers` ok.
- `ScrHammersleySearchPlusMiddlePoint` is excellent for super parallel cases (fully one-shot, i.e. `num_workers` = budget included) or for very multimodal cases (such as some of our MLDA problems); don't use softmax with this optimizer.
- `RandomSearch` is the classical random search baseline; don't use softmax with this optimizer.

In the long run we want to wrap all of them, **but if you are tackling this as your first issue you should focus on one. In that case please coment below which optimizer you are going to work on so we don't duplicate efforts.**

**Comment by AashifAmeer** [comment]
Hi @janosg , Under the Nevergrad library which optimization you would suggest me to include here, CMA-ES or Random search or PSO?

**Comment by janosg** [comment]
Hi @AashifAmeer, I you can use any optimizer. Start with the one that seems easiest to do. Long run we want all of them!

**Comment by janosg** [comment]
@AashifAmeer I extended the issue description a bit so more people can work simultaneously on the issue. Please comment below which optimizer you picked.

**Comment by gulshan-123** [comment]
Hi @janosg , I want to work on this issue. I will work on OnePlusOne algorithm. I have read the documentation from nevergrad. So, the nevergrad wrapper is expected to similar that of scipy algorithms, right?

**Comment by janosg** [comment]
This should help you to get started: https://optimagic.readthedocs.io/en/latest/how_to/how_to_add_optimizers.html

**Comment by gulshan-123** [comment]
I have written some basic structure of code and tested locally by installing nevergrad in optimagic conda environment.
But pre-commit is giving this error. I have added nevergrad to environments.yml too.
src/optimagic/optimizers/nevergrad_optimizers.py:7: error: Cannot find implementation or library stub for module named "nevergrad"  [import-not-found]

I am using ubuntu 24.04.

[https://github.com/PyCQA/docformatter] 
==> File /tmp/tmpivrx7hl6/.pre-commit-hooks.yaml
==> At Hook(id='docformatter-venv')
==> At key: language
=====> Expected one of conda, coursier, dart, docker, docker_image, dotnet, fail, golang, haskell, julia, lua, node, perl, pygrep, python, r, ruby, rust, script, swift, system but got: 'python_venv'

**Comment by r3kste** [comment]
> PSO is excellent in terms of robustness, high num_workers ok.

Hi. I have opened #579 which implements a wrapper for the PSO algorithm.

**Comment by r3kste** [comment]
Hello @janosg. Thanks for the feedback.

> we currently have the problem that they way you implement parallelism is not compatible with our history collection.

I want to point out that in #579, I actually set `disable_history=True` for PSO. This is because PSO is a swarm algorithm, and it seems like internally `nevergrad` calls the objective for each particle in the population. Due to this, I believe that maintaining history for PSO doesn't make much sense, similar to the case of other optimizers like `scipy_brute` and `scipy_differential_evolution`.

As `disable_history=True`, I believe there is no need of the custom Executor implementation. I would love to hear your thoughts on this.

**Comment by janosg** [comment]
We actually want to collect histories for all optimizers. The main benefit of history collection is to compare how efficient different optimizers are, i.e. how fast (in terms of function evaluations) they make progress.  

There are a few optimizers where we have not found ways to do it yet, but in nevergrad it seems to be quite easy, so we definitely want to do it.

**Comment by gulshan-123** [comment]
![Image](https://github.com/user-attachments/assets/70c34a62-4539-4774-8b89-6258cd63c321)

Hi @janosg,

I have gone through the minimize function in the Optimizer class of Nevergrad ([link](https://github.com/facebookresearch/nevergrad/blob/main/nevergrad/optimization/base.py#L717)). From my understanding, to enable parallelism, we need to implement a custom executor, as you had suggested.

However, I noticed that since minimize performs optimization by "asking" for one point at a time (line 723, 728), it might be more appropriate to internally call fun instead of batch_fun.

If we wish to utilize batch_fun, we would likely need to rely on the ask and tell interface to collect multiple recommendations and then evaluate them in batch (previously suggested approach).

**Comment by janosg** [comment]
@gulshan-123 Isn't the [`batch_mode`](https://facebookresearch.github.io/nevergrad/optimization.html#using-several-workers) argument there to solve exactly this problem? 

And no, you cannot just put fun into a ThreadPoolExecutor as history collection would not work if fun is called in multiple processes. That's the reason why `batch_fun` exists.

**Comment by r3kste** [comment]
Hello. Chiming in to hopefully clear up any confusion.

> Isn't the `batch_mode` argument there to solve exactly this problem?

With the way that the `minimize()` function is implemented in nevergrad, **regardless** of `batch_mode` it always 'submits' **exactly one point** to the executor to run the `objective_function()` on.

> With `batch_mode=True` it will ask the optimizer for `num_workers` points to evaluate, run the evaluations, then update the optimizer with the `num_workers` function outputs, and repeat until the budget is all spent.

This statement is correct, but the issue here is that it doesn't actually 'collect' the points to send it as a batch, rather it submits each point (from the batch of `num_workers` points) **one-by-one**.

### Solution

Due to the above mentioned reasons, I believe we cannot use the `minimize()` interface of the optimizer. Instead, we could use the `ask()` and `tell()` interface of the optimizer:

```python
while optimizer.num_ask < optimizer.budget:
    x_list = [optimizer.ask() for _ in range(optimizer.num_workers)]
    losses = problem.batch_fun(
        [x.value[0][0] for x in x_list], n_cores=self.n_cores
    )
    for x, loss in zip(x_list, losses, strict=True):
        optimizer.tell(x, loss)

recommendation = optimizer.provide_recommendation()
```

I would like to know your thoughts on this, and whether I missed out on anything.

**Comment by janosg** [comment]
@r3kste thanks for the clarification. 

Sure, if we cannot make it work with `minimize` we can use the ask-and-tell interface. Long run that is probably also the better choice if we want to have a more sophisticated outer loop.

**Comment by gulshan-123** [comment]
> range(optimizer.num_workers)

I think here we should have `range(min(optimizer.budget-optimizer.num_ask, optimizer.num_workers))`, for strict budget constraint.

**Comment by gulshan-123** [comment]
Hi @janosg, I think you have mistakenly closed this issue.
